File: leetcode/python/000141.linked-list-cycle.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
	# Time complexity: O(n)
	# Space complexity: O(1)
	def hasCycle(self, head: Optional[ListNode]) -> bool:
		if not head or not head.next:
			return False

		slow, fast = head, head
		while slow and fast and fast.next:
			slow = slow.next
			fast = fast.next.next
			if slow == fast:
				return True

		return False

File: leetcode/typescript/000141.linked-list-cycle.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function hasCycle(head: ListNode | null): boolean {
  const set = new Set<ListNode>();
  let cur = head;

  while (cur) {
    if (set.has(cur)) {
      return true;
    }

    set.add(cur);
    cur = cur.next;
  }

  return false;
}
// Time complexity: O(n)
// Space complexity: O(n)

function hasCycle2(head: ListNode | null): boolean {
  let slow = head;
  let fast = head;

  while (fast && fast.next) {
    fast = fast.next.next;
    slow = slow!.next;

    if (fast === slow) return true;
  }

  return false;
}
// Time complexity: O(n)
// Space complexity: O(1)
